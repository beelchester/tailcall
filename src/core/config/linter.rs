use std::collections::BTreeSet;

use anyhow::Result;
use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};

use super::Config;
use crate::core::config::Variant;
use crate::core::macros::MergeRight;
use crate::core::valid::{Valid, Validator};

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, schemars::JsonSchema, MergeRight)]
pub enum TextCase {
    #[serde(rename = "camelCase")]
    CamelCase,
    #[serde(rename = "pascalCase")]
    PascalCase,
    #[serde(rename = "snakeCase")]
    SnakeCase,
    #[serde(rename = "screamingSnakeCase")]
    ScreamingSnakeCase,
    #[serde(rename = "allCaps")]
    AllCaps,
}

impl From<TextCase> for Case {
    fn from(text_case: TextCase) -> Self {
        match text_case {
            TextCase::CamelCase => Case::Camel,
            TextCase::PascalCase => Case::Pascal,
            TextCase::SnakeCase => Case::Snake,
            TextCase::ScreamingSnakeCase => Case::ScreamingSnake,
            TextCase::AllCaps => Case::Upper,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, schemars::JsonSchema, MergeRight)]
pub struct Linter {
    pub default: Option<bool>,
    #[serde(rename = "autoFix")]
    pub autofix: Option<bool>,
    #[serde(rename = "field")]
    pub field_lint: Option<TextCase>,
    #[serde(rename = "type")]
    pub type_lint: Option<TextCase>,
    #[serde(rename = "enum")]
    pub enum_lint: Option<TextCase>,
    #[serde(rename = "enumValue")]
    pub enum_value_lint: Option<TextCase>,
}

pub fn lint(mut config: Config) -> Result<Config> {
    let lint = config.clone().server.lint;
    if let Some(lint) = lint {
        let autofix = lint.autofix.unwrap_or(false);
        if lint.field_lint.is_some() || lint.default.unwrap_or(false) {
            config = field_lint(config, autofix).to_result()?;
        }
        if lint.type_lint.is_some() || lint.default.unwrap_or(false) {
            config = type_lint(config, autofix).to_result()?;
        }
        if lint.enum_lint.is_some() || lint.default.unwrap_or(false) {
            config = enum_lint(config, autofix).to_result()?;
        }
        if lint.enum_value_lint.is_some() || lint.default.unwrap_or(false) {
            config = enum_value_lint(config, autofix).to_result()?;
        }
    }
    Ok(config)
}

pub fn field_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.field_lint.unwrap_or(TextCase::CamelCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.types.iter_mut().for_each(|(_, type_)| {
            for field in type_.fields.clone() {
                let case = field.0.to_case(target_case);
                if case != *field.0 {
                    if autofix {
                        if let Some(field_info) = type_.fields.remove(&field.0) {
                            tracing::warn!(
                                "field {} is renamed to {}",
                                field.0,
                                field.0.to_case(target_case)
                            );
                            type_
                                .fields
                                .insert(field.0.to_case(target_case), field_info);
                        }
                    } else {
                        errors.push(format!(
                            "lint failed for field {}, expected {}",
                            field.0, case
                        ));
                    }
                }
            }
        });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}

pub fn type_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.type_lint.unwrap_or(TextCase::PascalCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.clone().types.iter_mut().for_each(|(name, _)| {
            if name.to_case(target_case) != *name {
                if autofix {
                    if let Some(type_info) = config.types.remove(name.as_str()) {
                        tracing::warn!("type {} is renamed to {}", name, name.to_case(target_case));
                        config.types.insert(name.to_case(target_case), type_info);
                    }
                } else {
                    errors.push(format!(
                        "lint failed for type {}, expected {}",
                        name,
                        name.to_case(target_case)
                    ));
                }
            }
        });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}

pub fn enum_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.enum_lint.unwrap_or(TextCase::PascalCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.clone().enums.iter_mut().for_each(|(name, _)| {
            if name.to_case(target_case) != *name {
                if autofix {
                    if let Some(enum_info) = config.enums.remove(name.as_str()) {
                        tracing::warn!("enum {} is renamed to {}", name, name.to_case(target_case));
                        config.enums.insert(name.to_case(target_case), enum_info);
                    }
                } else {
                    errors.push(format!(
                        "lint failed for enum {}, expected {}",
                        name,
                        name.to_case(target_case)
                    ));
                }
            }
        });
    }
    Valid::succeed(config)
}

pub fn enum_value_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.enum_value_lint.unwrap_or(TextCase::AllCaps);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config
            .clone()
            .enums
            .iter_mut()
            .for_each(|(enum_name, enum_)| {
                let updated_variants: BTreeSet<_> = enum_
                    .variants
                    .iter()
                    .map(|variant| {
                        let case_name = variant.name.to_case(target_case);
                        if case_name != variant.name {
                            if autofix {
                                tracing::warn!(
                                    "variant {} is renamed to {}",
                                    variant.name,
                                    case_name
                                );
                                // return the variant with the updated name
                                Variant { name: case_name, ..variant.clone() }
                            } else {
                                errors.push(format!(
                                    "lint failed for enum variant {}, expected {}",
                                    variant.name, case_name
                                ));
                                variant.clone()
                            }
                        } else {
                            variant.clone()
                        }
                    })
                    .collect();

                config.enums.remove(enum_name);
                enum_.variants = updated_variants;
                config.enums.insert(enum_name.into(), enum_.clone());
            });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}
