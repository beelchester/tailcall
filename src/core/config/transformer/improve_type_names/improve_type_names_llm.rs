use std::collections::HashMap;

use reqwest;
use reqwest::Error;

use crate::core::config::Config;

pub struct ImproveTypeNamesLLM;

type OriginalTypeName = String;
type AIGeneratedTypeName = String;

struct LLMRequest {
    original_type_name: OriginalTypeName,
    prompt: String,
}

struct LLMResponse {
    original_type_name: OriginalTypeName,
    suggested_type_names: Vec<AIGeneratedTypeName>,
}

impl ImproveTypeNamesLLM {
    pub fn generate_llm_type_names(
        config: Config,
    ) -> Result<HashMap<OriginalTypeName, Vec<AIGeneratedTypeName>>, String> {
        let mut llm_requests: Vec<LLMRequest> = Vec::new();

        for (type_name, type_info) in config.types.iter() {
            if config.is_root_operation_type(type_name) {
                continue;
            }

            let mut fields: Vec<String> = Vec::new();
            for (field_name, field_info) in type_info.fields.iter() {
                fields.push(format!("{}: {}", field_name, field_info.type_of));
                //ex: T2: {name: String, age: Int}
            }

            let prompt = Self::generate_llm_request(type_name, fields);
            llm_requests.push(prompt);
        }

        #[cfg(not(target_arch = "wasm32"))]
        return tokio::task::block_in_place(move || Self::make_llm_requests(llm_requests));
        #[cfg(target_arch = "wasm32")]
        Self::make_llm_requests(llm_requests)
    }

    fn generate_llm_request(name: &str, fields: Vec<String>) -> LLMRequest {
        let mut type_prompt = format!("{} {{", name);

        for (index, field) in fields.iter().enumerate() {
            type_prompt.push_str(field);
            if index != fields.len() - 1 {
                type_prompt.push(' ');
            }
        }
        type_prompt.push('}');
        let prompt = format!("{{\"prompt\":\"{}\"}}", type_prompt);
        LLMRequest { original_type_name: name.to_string(), prompt }
    }

    fn make_llm_requests(
        llm_requests: Vec<LLMRequest>,
    ) -> Result<HashMap<OriginalTypeName, Vec<AIGeneratedTypeName>>, String> {
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .unwrap();
        runtime.block_on(async {
            let client = reqwest::Client::new();

            let mut results: HashMap<OriginalTypeName, Vec<AIGeneratedTypeName>> = HashMap::new();
            for llm_request in llm_requests.into_iter() {
                let client = client.clone();
                let response = Self::get_llm_response(client, llm_request).await;
                match response {
                    Ok(result) => {
                        results.insert(result.original_type_name, result.suggested_type_names);
                    }
                    Err(e) => return Err(e.to_string()),
                };
            }

            Ok(results)
        })
    }

    async fn get_llm_response(
        client: reqwest::Client,
        llm_request: LLMRequest,
    ) -> Result<LLMResponse, Error> {
        let response = client
            .post("https://ec56-2405-201-101e-60f3-fc34-439-491b-fe01.ngrok-free.app/type_name")
            .header("Content-Type", "application/json")
            .header(reqwest::header::TRANSFER_ENCODING, "chunked")
            .body(llm_request.prompt)
            .send()
            .await?;

        let result = response.text().await?;
        let result = result.replace([' ', '\n', '[', ']', '"'], ""); //ex: [User,Profile,Contact,Person,Entity]
        let result = result.split(',').map(|s| s.to_string()).collect();
        Ok(LLMResponse {
            original_type_name: llm_request.original_type_name.to_string(),
            suggested_type_names: result,
        })
    }
}
