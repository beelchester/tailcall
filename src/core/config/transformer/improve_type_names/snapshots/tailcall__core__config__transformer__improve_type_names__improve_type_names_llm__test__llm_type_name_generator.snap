---
source: src/core/config/transformer/improve_type_names/improve_type_names_llm.rs
expression: "format!(\"{:?}\", llm_response)"
---
{"T3": ["Color", "Entry", "Palette", "Shade", "Hue"], "T1": ["ColorItem", "Hue", "Palette", "Tint", "Shade"], "T2": ["Palette", "Theme", "Scheme", "Design", "Style"]}
